ncm-metaconfig/src/main/perl/metaconfig.pm

```diff
@@ -265,55 +329,89 @@ sub handle_service
         return;
     }
 
+    if (! $self->run_shell_command($commands, 'test', "$fh")) {
+        $fh->cancel();
+        return;
+    };
+
     if ($fh->close()) {
         $self->info("File $file updated");
-        $self->prepare_action($srv, $file, $actions);
+        $self->prepare_daemon_action($srv, $file, $daemon_actions);
+        return if ! $self->run_shell_command($commands, 'changed');
     } else {
         $self->verbose("File $file up-to-date");
     };
 
-    return 1;
+    return $self->run_shell_command($commands, 'post');
 }
 
+# Lookup actions in command registry, and return hashref with actual commands for each action
+sub resolve_command_actions
+{
+    my ($self, $command_registry, $actions) = @_;
+
+    my $commands = {};  # this will trigger reporting that nothing is configured is this stays empty
+    foreach my $type (sort keys %$actions) {
+        my $action = $actions->{$type};
+        my $command = $command_registry->{$action};
+        if ($command) {
+            $commands->{$type} = $command;
+            $self->verbose("Resolved $type action $action to command '$command'");
+        } else {
+            # Not fatal, should be covered in schema already
```

Comment:
What is meant by "not fatal" here? In some cases (i.e. for dependent components, any error() is "fatal". The rest of the comment reads as, "Should not happen due to this being validated in the schema"?

Reply:
fatal as in stop parsing and do not execute the rest of the component code. thinking about it, i'd better make this fatal as it is the right thing to do, and the schema prevents that you can use such a config in a profile.

Reply:
Seems harsh (since in theory the code could still deliver some useful value) but probably makes sense given it's a schema check.

Reply:
i guess it's a personal preference. i prefer things stricter, i guess from your experience some useful value is probably better than nothing at all.
if you come up with better comment, i'll leave the code non-fatal ;)

Reply:
```suggestion
            # Should not happen due to this being validated in the schema
```

Reply:
Re-upping my suggested wording change.